Source/WebCore/dom/StaticRange.cpp

```diff
@@ -83,4 +83,23 @@ void StaticRange::visitNodesConcurrently(JSC::AbstractSlotVisitor& visitor) cons
     addWebCoreOpaqueRoot(visitor, end.container.get());
 }
 
+bool StaticRange::isValid() const
+{
+    Node& startContainer = this->startContainer();
+    Node& endContainer = this->endContainer();
+
+    if (collapsed())
+        return false;
+    if (endOffset() < startOffset() && startContainer.isSameNode(&endContainer))
+        return false;
+    if (startContainer.isDocumentNode() || startContainer.isDocumentTypeNode())
```

Comment:
Where did this check come from. It's not listed in https://dom.spec.whatwg.org/#staticrange-valid.

Reply:
There is a WPT `imported/w3c/web-platform-tests/css/css-highlight-api/painting/custom-highlight-painting-staticrange-001.html`. One case has `document` as the container and it is deemed invalid since the first child of `document` would be `<!DOCTYPE html>` which doesn't make sense to style. So nothing will/should be painted.

Reply:
Have you considered the possibility that WPT is wrong?

https://dom.spec.whatwg.org/#staticrange-valid doesn't check this condition and there is nothing regarding document node or document type node in https://drafts.csswg.org/css-highlight-api-1/#range-invalidation

Either we're missing some text in the spec, which should be filed as a spec issue, or that particular test is wrong.

Reply:
I asked people from web standards and they also said it could potentially be wrong but the logic would be to grab `<!DOCTYPE html>` in which you can't highlight that. Thus, don't paint. They said we could discuss it more to determine if the test is wrong but so far Chrome and Firefox both pass this test as well.

Reply:
Not painting makes sense. What doesn't is making this isValid concept depart from the definition in the spec.

I think we should get rid of these checks from isValid and placed them in where isValid is called instead.

Alternatively, we should add a check comment saying this doesn't match the spec's definition of isValid and add a URL to a spec issue.

Reply:
Put it differently, conceptually, there is a concept of static range being valid vs. static range being paintable as highlights. It totally makes sense that a static range anchored at doctype can't be painted but that's a different concept from static range being valid, which has a very specific definition.

Also, why is anchoring node on document node wrong / invalid? If we specified `display: block` on `<html>` element, it will be rendered in which case highlight would be paintable at least in concept.

Reply:
By getting rid of the checks, did you mean the ones not explicitly stated in the spec or just get rid of the function? Also, for the case you mentioned, I guess it _would_ be paintable but I don't think it _should_? I wouldn't know what the styling would apply to? I'll ask to make sure of this.